File: 백준/Gold/20040. 사이클 게임/사이클 게임.js
const fs = require("fs");
const input = fs.readFileSync("/dev/stdin").toString().trim().split("\n");

const [[n, m], ...vertex] = input.map((str) => str.split(" ").map(Number));

function solution(n, vertex) {
  const root = Array.from({ length: n }, (_, idx) => idx);

  const find = (a) => (root[a] === a ? a : (root[a] = find(root[a])));

  const union = (a, b) => {
    const rootA = find(a);
    const rootB = find(b);

    if (rootA < rootB) {
      root[rootB] = rootA;
    } else {
      root[rootA] = rootB;
    }
    return rootA !== rootB;
  };

  return vertex.reduce(
    (answer, [a, b], idx) => (answer === 0 && !union(a, b) ? idx + 1 : answer),
    0
  );
}

console.log(solution(n, vertex));

File: 백준/Gold/20040. 사이클 게임/사이클 게임.cc
#include <iostream>
#define MAX 500000
using namespace std;

int root[MAX] = {0, };

int find(int a);
bool unionTwo(int a, int b);

int main() {
  ios_base::sync_with_stdio(false);
  cin.tie(NULL);
  int n, m;
  cin>>n>>m;

  for(int i = 0; i < n; i++) {
    root[i] = i;
  }

  int a, b, findCycle = 0;
  for(int i = 1; i <= m; i++) {
    cin>>a>>b;
    findCycle = !unionTwo(a, b) && findCycle == 0 ? i : findCycle;
  }
  cout<<findCycle<<'\n';
  return 0;
}

int find(int a) {
  return root[a] == a ? a : root[a] = find(root[a]);
}

bool unionTwo(int a, int b) {
  int rootA = find(a);
  int rootB = find(b);
  root[rootA] = rootA > rootB ? rootB : rootA;
  root[rootB] = rootA > rootB ? rootB : rootA;
  return rootA != rootB;
}